src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -507,34 +520,50 @@ fn resolution_failure(
     let sp = span_of_attrs(attrs);
     let msg = format!("`[{}]` cannot be resolved, ignoring it...", path_str);
 
-    let code_dox = sp.to_src(cx);
-
-    let doc_comment_padding = 3;
     let mut diag = if let Some(link_range) = link_range {
-        // blah blah blah\nblah\nblah [blah] blah blah\nblah blah
-        //                       ^    ~~~~~~
-        //                       |    link_range
-        //                       last_new_line_offset
-
         let mut diag;
-        if dox.lines().count() == code_dox.lines().count() {
-            let line_offset = dox[..link_range.start].lines().count();
-            // The span starts in the `///`, so we don't have to account for the leading whitespace
-            let code_dox_len = if line_offset <= 1 {
-                doc_comment_padding
-            } else {
-                // The first `///`
-                doc_comment_padding +
-                    // Each subsequent leading whitespace and `///`
-                    code_dox.lines().skip(1).take(line_offset - 1).fold(0, |sum, line| {
-                        sum + doc_comment_padding + line.len() - line.trim_start().len()
-                    })
-            };
 
-            // Extract the specific span
+        if attrs.doc_strings.iter().all(|frag| match frag {
+            DocFragment::SugaredDoc(..) => true,
+            _ => false,
+        }) {
+            let source_dox = sp.to_src(cx);
+            let mut source_lines = source_dox.lines().peekable();
+            let mut md_lines = dox.lines().peekable();
+
+            // The number of bytes from the start of the source span to the resolution failure that
+            // are *not* part of the markdown, like comment markers.
+            let mut source_offset = 0;
+
+            // Eat any source lines before the markdown starts (e.g., `/**` on its own line).
+            while let Some(source_line) = source_lines.peek() {
+                if source_line.contains(md_lines.peek().unwrap()) {
+                    break;
+                }
+
+                // Include the newline.
+                source_offset += source_line.len() + 1;
+                source_lines.next().unwrap();
+            }
+
+            // The number of lines up to and including the resolution failure.
+            let num_lines = dox[..link_range.start].lines().count();
+
+            // Consume inner comment markers (e.g., `///` or ` *`).
+            for (source_line, md_line) in source_lines.zip(md_lines).take(num_lines) {
+                source_offset += if md_line.is_empty() {
+                    // If there is no markdown on this line, then the whole line is a comment
+                    // marker. We don't have to count the newline here because it's in the markdown
+                    // too.
+                    source_line.len()
+                } else {
+                    source_line.find(md_line).unwrap()
```

Comment:
This can panic on docs like https://github.com/rust-lang/rust/blob/b866f7d258a2428e004039eb0f3fabd73cc9d6ae/src/librustc_mir/util/elaborate_drops.rs#L558-L573 because of the blank line. This is what caused #56531 to fail.

@bors r-